### PR TITLE
[Azure Pipelines] Use Python 3.8.x for Edge test-suite runs

### PR DESCRIPTION
**Pull request by stephenmcgruer**
See the RFC: https://github.com/web-platform-tests/rfcs/pull/65

This also affects `infrastructure/ tests: Windows 10`, but that job is currently only manually triggered and is basically unused.

### CODE REVIEW

**Comment by stephenmcgruer** [comment]
cc @ziransun 

https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=52438&view=results should be an Edge Dev run with this commit.

File: .azure-pipelines.yml
Comment (stephenmcgruer) [inline .azure-pipelines.yml:490]: @foolip - can you educate me a bit on the 'Hosted Windows Client' pool? I initially tried to use it, but found that it didn't have py3 available and I wasn't sure how to make it available (UsePythonVersion isn't a task there, and `tools/ci/azure/install_python.yml` seems to *manually* download py2.7 and install it? :O). I tried to poke around azure pipelines, but I don't have access to the web-platform-tests organization - I only have access to the wpt pipeline itself.

Comment (foolip) [reply .azure-pipelines.yml:490]: This pool is maintained by @mustjab and updated weekly by a process that's opaque to us (outside Microsoft) but has worked pretty reliably so far. To get additional software installed by default, @mustjab would have to help.

If we find that windows-2019 works to run the tests now, we should actually switch to that. When we set this up, none of the agents available by default could launch Edge classic. Perhaps now it's possible to run the new Edge in one of these VMs.

I'll try to add you to the web-platform-tests org in Azure DevOps.

Comment (mustjab) [reply .azure-pipelines.yml:490]: You should be able to update install_python.yml to add a task to download and install py3. Do we want to install specific py3 version? Do you have a link to py3 MSI installer?

**Comment by stephenmcgruer** [comment]
The Edge run finished but there was an error (I suspect py3-caused) that resulted in it getting no 'revision' tag, which means wpt.fyi won't process it. Unfortunately such errors are deliberately swallowed. I've uploaded a patch that will hopefully log the error, and set off another run - https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=52472&view=results

**Comment by stephenmcgruer** [comment]
Yep, vcs.py grabs the output of `git rev-parse --show-cdup` via `subprocess.check_output`, then compares it to `"\n"`, which fails since the actual output is `b"\n"`. I don't understand yet why this appears to work on Linux on Py3 (given our previous success wpt.fyi uploads), but for now I'm going to just stick a fix in here and trigger another Edge Dev run: https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=52525&view=results

**Comment by stephenmcgruer** [comment]
> I don't understand yet why this appears to work on Linux on Py3 (given our previous success wpt.fyi uploads)

https://community-tc.services.mozilla.com/tasks/f0VA2hcISQClpnai1ahO-w/runs/0/logs/live/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2Ff0VA2hcISQClpnai1ahO-w%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log appears to confirm that it **shouldn't**, but yet we've definitely had successful uploads from Linux+Py3 in the past. Once that run is done I'll pull down both the artifacts for it (to see the run_info) and look at the wpt.fyi logs (to see if and how it passes the processor).

**Comment by stephenmcgruer** [comment]
Moving back to Windows, the latest Edge Dev Py 3.8.5 run fails at the end with:

```
Traceback (most recent call last):
  File "D:\a\1\s\_venv3\Lib\site-packages\mozlog\structuredlog.py", line 274, in _handle_log
    handler(data)
  File "D:\a\1\s\_venv3\Lib\site-packages\mozlog\handlers\base.py", line 74, in __call__
    formatted = self.formatter(data)
  File "D:\a\1\s\_venv3\Lib\site-packages\mozlog\handlers\base.py", line 42, in __call__
    return self.inner(item)
  File "D:\a\1\s\_venv3\Lib\site-packages\mozlog\reader.py", line 74, in __call__
    return handler(data)
  File "D:\a\1\s\tools\wptrunner\wptrunner\formatters\wptreport.py", line 73, in suite_end
    return json.dumps(self.results) + "\n"
  File "C:\hostedtoolcache\windows\Python\3.8.5\x64\lib\json\__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "C:\hostedtoolcache\windows\Python\3.8.5\x64\lib\json\encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "C:\hostedtoolcache\windows\Python\3.8.5\x64\lib\json\encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "C:\hostedtoolcache\windows\Python\3.8.5\x64\lib\json\encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable
```

So there are some bytes somewhere in self.results I guess.

**Comment by stephenmcgruer** [comment]
Success, we have Py3.8.5 + Edge Dev results! https://wpt.fyi/results/?run_id=640430002

https://wpt.fyi/results/?diff&filter=ADC&run_id=626930002&run_id=640430002 is the diff to the nearest run (2df9cdd has two commits that aren't in the RHS, but should be minor). FYI @ziransun .

File: .azure-pipelines.yml
Comment (stephenmcgruer) [reply .azure-pipelines.yml:490]: @mustjab - I wasn't able to locate an amd64 msi at https://www.python.org/ftp/python/3.8.5/ . There's a subdirectory, https://www.python.org/ftp/python/3.8.5/amd64/, with **many** msi files - but not sure which (if any) are correct! :/

Comment (mustjab) [reply .azure-pipelines.yml:490]: Looks like they ship them as .exe files now, so a link like this would work: https://www.python.org/ftp/python/3.7.9/python-3.7.9-amd64.exe and you can use this guide on how to run this installer without UI: https://docs.python.org/3.7/using/windows.html

**Comment by stephenmcgruer** [comment]
Doing another Edge run is blocked on https://github.com/web-platform-tests/wpt/issues/25923

**Comment by stephenmcgruer** [comment]
@mustjab I cannot figure out what is going on here. It seems like for 2/3rds of the bots, the install of python just silently fails :/.

Compare two bots from the same run:
  https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=55642&view=logs&jobId=03cdc71c-1e73-5e64-c319-92146034426c&j=03cdc71c-1e73-5e64-c319-92146034426c&t=8dca10db-2bd8-55e7-a034-c365bf0091cd
  https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=55642&view=logs&jobId=03cdc71c-1e73-5e64-c319-92146034426c&j=2b376673-b6b7-55d3-67e2-e8305abf5c1e&t=a6e1fae8-d0cc-5771-a2ef-9ba27692873c

On the first one, there just isn't a Python38 directory after running the installer, with no clear failed output. Am I doing anything obviously wrong here?

**Comment by mustjab** [comment]
We'll need to collect setup logs to figure out why python38 installation is failing. What is your current install command line?

**Comment by stephenmcgruer** [comment]
See the PR diff: https://github.com/web-platform-tests/wpt/pull/24952/files#diff-5fc6a428ee6444bed60ed4bb5ad58123

`    & .\python_installer.exe /passive TargetDir=C:\Python38`

**Comment by mustjab** [comment]
Let's add /log <filename> and upload that file as artifact of the install step or add Get-Content -Path <filename> to print content of that file

**Comment by stephenmcgruer** [comment]
Hah, and of course now all shards pass when I get the logging in place!

I'll set off another run later to try and reproduce and dig in, but for now @ziransun there is a diff at https://wpt.fyi/results/?diff&filter=ADC&run_id=714260003&run_id=704670001 which should unblock your work. (Sorry for the delay).

**Comment by ziransun** [comment]
Thanks @stephenmcgruer! I'll look into the new run result.

**Comment by stephenmcgruer** [comment]
And we have a failing log (cc @mustjab ) , excellent!

[Failing log](https://dev.azure.com/web-platform-tests/b14026b4-9423-4454-858f-bf76cf6d1faa/_apis/build/builds/55761/logs/88)

[Succeeding log](https://dev.azure.com/web-platform-tests/b14026b4-9423-4454-858f-bf76cf6d1faa/_apis/build/builds/55761/logs/57)

The big obvious different I see is that it looks like the failing log is trying to **install** Python3, whilst the succeeding log is trying to **Update** it?! (I have no idea why yet).

```
2020-10-07T20:04:43.1359044Z [1B44:0538][2020-10-07T13:04:42]i000: Setting string variable 'ActionLikeInstalling' to value 'Installing'
2020-10-07T20:04:43.1360385Z [1B44:0538][2020-10-07T13:04:42]i000: Setting string variable 'ActionLikeInstallation' to value 'Setup'
2020-10-07T20:04:43.1362373Z [1B44:1C80][2020-10-07T13:04:42]i200: Plan begin, 52 packages, action: Install

vs

2020-10-07T19:08:39.5922163Z [1BE8:1B88][2020-10-07T12:08:38]i000: Setting string variable 'ActionLikeInstalling' to value 'Updating'
2020-10-07T19:08:39.5924357Z [1BE8:1B88][2020-10-07T12:08:38]i000: Setting string variable 'ActionLikeInstallation' to value 'Modify'
2020-10-07T19:08:39.5925373Z [1BE8:09F0][2020-10-07T12:08:38]i200: Plan begin, 52 packages, action: Modify
```

The (failing) install one also ends with:

```
2020-10-07T20:04:43.1742481Z [1B44:1C80][2020-10-07T13:04:42]i299: Plan complete, result: 0x0
2020-10-07T20:04:43.1744117Z [1B44:1C80][2020-10-07T13:04:42]i300: Apply begin
2020-10-07T20:04:43.1745662Z [1B44:1C80][2020-10-07T13:04:42]i010: Launching elevated engine process.
2020-10-07T20:04:43.1747347Z [1B44:1C80][2020-10-07T13:04:42]i011: Launched elevated engine process.
```

And then nothing.

**Comment by stephenmcgruer** [comment]
@mustjab - do you have any thoughts on the comment above (https://github.com/web-platform-tests/wpt/pull/24952#issuecomment-705512446) ?

**Comment by mustjab** [comment]
I see both installers start with this:

Initializing string variable 'ActionLikeInstalling' to value 'Installing'
Initializing string variable 'ActionLikeInstallation' to value 'Setup'

But in passing case the subsequent install action is "update" while failing one remains in "install" state. Do you know what is the installer exit code? I don't see one logged in the failing case. You can print exit code using $LastExitCode after running python installer.

**Comment by stephenmcgruer** [comment]
@mustjab thanks. Can you take a quick look at the script (https://github.com/web-platform-tests/wpt/pull/24952/files) again? I added:

`    Write-Host "Exit code was: $LastExitCode"`

But in the [current run](https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=56527&view=logs&j=2b376673-b6b7-55d3-67e2-e8305abf5c1e&t=a6e1fae8-d0cc-5771-a2ef-9ba27692873c) (which is succeeding) this just prints out:

`Exit code was: `

**Comment by mustjab** [comment]
The error logging code looks right to me. The installer generates a bunch of log_*.txt files, for each component that it installs. I tried running your script on a local VM and was able to repro the install failure with LastExitCode =1

Looking at the generated log files, I do see these error messages:
  log.txt:383:[1B20:2300][2020-10-26T11:58:42]w363: Could not create system restore point, error: 0x80070422. Continuing...
  log.txt:420:[2610:15DC][2020-10-26T11:58:57]e000: Error 0x80070643: Failed to install MSI package.
  log.txt:421:[2610:15DC][2020-10-26T11:58:57]e000: Error 0x80070643: Failed to configure per-user MSI package.
  log.txt:423:[2610:15DC][2020-10-26T11:58:57]e000: Error 0x80070643: Failed to execute MSI package.

In my case the install failure was due to lack of disk space on the VM, but there was no specific error around that, instead we get error 0x80070643 = ERROR_INSTALL_FAILURE

Do we have more failure logs?

**Comment by stephenmcgruer** [comment]
I spent some more time trying to debug this via bot today, but made little to no progress. You can check out my latest scripts in this PR. I did manage to get the log files published, but I think that they're overwriting each other (all named `log.txt` for the 10 Azure Dev jobs). It also seems like bots are reused without any cleanup done, so its really hard to figure out what the status is - I'm seeing runs where `C:\Python38` already exists, and where there are already `log.txt` files *in the WPT checkout*... I ended up adding more code to clean that up.

I could definitely imagine that we might have bad state from earlier bots sticking around and thus causing problems installing Python. For example, even on the bots that already have Py38 installed and so pass the running of `python_installer.exe`, the next step calling `pip --disable-pip-version-check install --upgrade virtualenv` often fails with no output ([example](https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=56923&view=logs&jobId=2b376673-b6b7-55d3-67e2-e8305abf5c1e&j=25a61619-8fbe-599f-66e7-d151be880261&t=208720e2-82d0-5012-9bab-e5c2594b7729)). I could totally believe that that is due to some weird corruption/etc.

@mustjab - how do we get these bots back to a clean state? Or, if they are meant to be reused between runs, shouldn't we install Py3.8 just once at initial bot startup rather than as part of each run (and how can we do that?). FYI this problem installing Python 3.8 is now blocking the WPT-goes-to-Py3 RFC (https://github.com/web-platform-tests/rfcs/pull/65) :(

**Comment by mustjab** [comment]
The agents do get re-used and we rebuild them once a week, every Sunday. To include python38 as part of the machine image, we'll have to modify the scripts that are used to create these agents. Let me take a look and see what we can do as I haven't touched these scripts in some time now. Which specific python 3.8 version do we want to have included?

**Comment by stephenmcgruer** [comment]
3.8.5 would be great. I'm also happy if we can find some way to better debug this problem and dig in - just feel quite useless atm :(

**Comment by mustjab** [comment]
@stephenmcgruer I've updated your install script, can you trigger a run to validate it?

**Comment by stephenmcgruer** [comment]
> @stephenmcgruer I've updated your install script, can you trigger a run to validate it?

Thanks, triggered one now!

**Comment by stephenmcgruer** [comment]
@mustjab Looks like the arguments to `-ArgumentList` were dropped. Did you intend to remove them (and thus we should remove `-ArgumentList`) or was that a mistake (and thus we should restore them)?

**Comment by mustjab** [comment]
That was mistake on my part, added them back now

**Comment by stephenmcgruer** [comment]
@mustjab - ok we have some more runs, sadly all failures. Looking at a few:

* Edge Dev 1 fails, but python_installer had exit code 0 ([logs](https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57003&view=logs&jobId=2b376673-b6b7-55d3-67e2-e8305abf5c1e&j=2b376673-b6b7-55d3-67e2-e8305abf5c1e&t=a6e1fae8-d0cc-5771-a2ef-9ba27692873c)), but there is no Python38 directory created.
    * I do see some errors in the [install log](https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57003&view=logs&jobId=2b376673-b6b7-55d3-67e2-e8305abf5c1e&j=2b376673-b6b7-55d3-67e2-e8305abf5c1e&t=57b916e7-31b3-59d4-7b57-37a252ccf828), but the ultimate exit code is 0
* Edge Dev 2 fails and python_installer had exit code 1603 ([logs](https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57003&view=logs&jobId=2b376673-b6b7-55d3-67e2-e8305abf5c1e&j=c47df06b-e4df-5c68-585c-fc35c97de295&t=8e99f000-757c-5ec9-56ea-3b255bfbb789)). 
   * From a brief google, this could either be "Windows Installer is attempting to install an app that is already installed on your PC." or "The SYSTEM account does not have Full Control permissions on the folder that you are trying to install the Windows Installer package to. You notice the error message because the Windows Installer service uses the SYSTEM account to install software." (or maybe something else?)
* Edge Dev 3, 5, 6, 7, 8, and 9 same as 2
* Edge Dev 4 and 10 same as 1

So at this point I think the bots are royally hosed 😆 . Is it possible to refresh them ahead of schedule, and then we can maybe try to start over and figure out the actual safe installation procedure? (I suspect I've caused all sorts of trouble).

It may be worth investigating if we can 'put aside' a single bot in a new pool or something where we can play with it and refresh it frequently until we figure out a workable path?

**Comment by mustjab** [comment]
I've rebuilt the agents, please give it another try and let's see how well installer works on clean images.

**Comment by stephenmcgruer** [comment]
Looks like the installer works fine on clean images. Once they've finished and are cycling back into the pool I'll do another run (e.g. tomorrow).

Btw, if you have write access (which I think you do since you can push to this branch), you can trigger runs yourself too! See https://web-platform-tests.org/running-tests/from-ci.html?highlight=trigger - the branch is `triggers/edge_dev`.

**Comment by stephenmcgruer** [comment]
Oh that's a fun one. In [this run](https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57042&view=logs&jobId=c47df06b-e4df-5c68-585c-fc35c97de295&j=c47df06b-e4df-5c68-585c-fc35c97de295&t=f296716c-d164-502b-304b-a9322a51fb6a) we delete both py27 and py38, then try to install with `/repair`, and we seem to recreate `py38` folder but not the scripts folder!?

More generally, looks like second run of the script on non-clean images did not work.

**Comment by mustjab** [comment]
Instead of deleting Python38 folder, i've updated the script to do uninstall. In my local testing that led to more consistent installer behavior.

**Comment by mustjab** [comment]
@stephenmcgruer Looks like the change to uninstall Python 3.8 instead of deleting python38 folder worked: https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57063&view=results

I had to rebuild the VMs again today as some of them were in a bad state, but feel free to re-run with latest changes and see if we can consistent install python.

**Comment by stephenmcgruer** [comment]
This is looking good! I've successfully run two runs, the second of which had to uninstall and reinstall py3.8 and did so fine. I think we're in a good place to turn this into a CL that we could land once we wanted to go to Py3.8 (e.g. by getting rid of most of the debug logging, etc, will do that today).

Thanks @mustjab for your work here - very much appreciated!

**Review by Hexcles — APPROVED** [review]
Rubber-stamp LGTM

I don't know much about Powershell but it seems like you've worked with an expert on this :)

**Pull request merged**
